File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet;
        for (string x : wordDict) wordSet.insert(x);
        int n = s.size();
        vector<string> dp(n + 1, "");
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i + 1; j <= n; j++) {
                string x = dp[j];
                string search = s.substr(i, (n - i - x.size()));
                if (wordSet.count(search)) {
                    dp[i] = s.substr(i);
                    break;
                }
            }
        }
        return dp[0] == s;
    }
};

File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        List<String> dp = new ArrayList<>();
        Set<String> wordSet = new HashSet<>(wordDict);
        for (int i = 0; i <= n; i++) {
            dp.add("");
        }
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i + 1; j <= n; j++) {
                String x = dp.get(j);
                if (wordSet.contains(s.substring(i, n - x.length()))) {
                    dp.set(i, s.substring(i, n));
                    break;
                }
            }
        }
        return dp.get(0).equals(s);
    }
}